### PR TITLE
Allow full control on Decimal behaviour

### LINKED ISSUE #90: Bypass number rounding
`babel.numbers.format_decimal()` and `babel.numbers.format_currency()` have a built-in banker rounding implemented. See: https://github.com/python-babel/babel/blob/3ec7bb104e808d1fda16fbf9fec4b9d7efccaa3f/babel/numbers.py#L648-L649

If we need to localize numbers without rounding, we can't use these methods.

The minimal default decimal precision per method are:

  * `babel.numbers.format_decimal()` => 3 digits after the dot
  * `babel.numbers.format_currency()` => 2 digits after the dot

Proof:

```python
>>> import babel
>>> set([babel.Locale.parse(l).decimal_formats._data[None].frac_prec
...      for l in babel.localedata.locale_identifiers()])
set([(0, 6), (0, 3)])
>>>
```

```python
>>> import babel
>>> set([babel.Locale.parse(l).currency_formats._data[None].frac_prec
...      for l in babel.localedata.locale_identifiers()])
set([(2, 2)])
>>>
```

So if you have monetary amounts to localize, and they're already rounded to 2 trailing digits, then you're lucky. Using any of the `format_*()` methods will have no side effects.

But for other numbers with higher precision, using `format_*()` as-is is dangerous. It will introduce unwanted rounding. For example, calling `format_currency(0.9999, 'EUR', locale='fr')` will return `1,00 €`. I expect here to get the pristine `0,9999 €` string.

I think there must be a clean and documented way to bypass arbitrary rounding when localizing numbers.

### PR DESCRIPTION
**Pull request by etanol**
It turns out that the solution to allow the user to change the rounding mode was way simpler than expected at the beginning.

According to [UTS #35][1] sections [3.3][2] and [3.7][3], the user is allowed to change rounding mode.  But the default should be _half even_ mode.  This is already the default Python rounding mode in all decimal implementations, so the solution was to stop indicating the rounding mode on `Decimal` operations.

Some documentation has been added to show how to properly change the rounding mode and other `Context` parameters.

Therefore, this should fix #90 

[1]: http://www.unicode.org/reports/tr35/tr35-numbers.html
[2]: http://www.unicode.org/reports/tr35/tr35-numbers.html#Formatting
[3]: http://www.unicode.org/reports/tr35/tr35-numbers.html#Rounding

### CODE REVIEW

File: babel/_compat.py [hunk 1]
```diff
@@ -58,19 +58,13 @@
...
-    Decimal = _cdec
-    InvalidOperation = (_invop, _cinvop)
-    ROUND_HALF_EVEN = _CRHE
-except ImportError:
-    Decimal = _dec
-    InvalidOperation = _invop
-    ROUND_HALF_EVEN = _RHE
+if sys.version_info[:2] >= (3, 3):
+    import decimal
+else:
```
Comment (etanol) [inline babel/_compat.py]: This seems odd, according to Travis, [`babel/_compat.py`][1] is 100% covered.

[1]: https://codecov.io/gh/python-babel/babel/src/f12e975b5af270b8846e8e01df37e3b3ad2e3b57/babel/_compat.py

**Comment by akx** [comment]
ack 6238f66 8341064 f12e975

**Comment by akx** [comment]
LGTM, but if someone else (@jtwang?) wants to give this an once-over, that'd be grand. :)

File: docs/numbers.rst [hunk 2]
```diff
@@ -86,6 +86,58 @@ the specification. The following table is just a relatively brief overview.
   +----------+-----------------------------------------------------------------+
 
 
+Rounding Modes
```
Comment (jtwang) [inline docs/numbers.rst]: +++++++

Thanks for the great docs!

Could you add a doctest or a note to the docstring for `babel.numbers.format_decimal`? Same for `format_currency` - especially around the interaction with custom currency formats and `currency_digits`. (omg so complicated)

File: docs/numbers.rst [hunk 3]
```diff
@@ -86,6 +86,58 @@ the specification. The following table is just a relatively brief overview.
...
+
+It is also possible to use ``decimal.setcontext`` or directly modifying the
+instance returned by ``decimal.getcontext``.  However, using a context manager
+is always more convenient due to the automatic restoration and the ability to
+nest them.
+
+Whatever mechanism is chosen, always make use of the ``decimal`` module imported
+from ``babel.numbers``.  For efficiency reasons, Babel uses the fastest decimal
+implementation available, such as `cdecimal`_.  These various implementation
+offer an identical API, but their types and instances do **not** interoperate
```
Comment (jtwang) [inline docs/numbers.rst]: Hmm... is this considered a non-backwards compatible change?

Comment (jtwang) [inline docs/numbers.rst]: Ohhh nevermind, we were already using `cdecimal`...

**Comment by jtwang** [comment]
Thank you so much for taking this on!

**Comment by sublee** [comment]
@jtwang No it doesn't.  What I need is the precision setting for `format_number` and the similar functions.  Not rounding option.

Precision to show usually depend on business logic, not specific locale.  The code for determining precision should be static against dynamic locales.  For example:

```python
>>> print(format_number(123456789.123456789, prec=5, locale=random.choice(locales)))
123.456.789,12346
```

For now, we can't separate precision from locales.  Because `format_number` detects precision from a `format` option but the format string comes from locales.

File: docs/numbers.rst (hunk 3 above)
Comment (etanol) [inline docs/numbers.rst]: Well, in terms of Babel API is fully backwards compatible.  If your code manipulates the current decimal context while also using Babel, you may see slightly different results.

But if you are playing with decimal precision and exponents, aside from the rounding mode, you should have seen this for some time already

**Comment by etanol** [comment]
The rounding performed by babel is just for the purpose of digit truncation before formatting a numeric or monetary amount.  That said, you can always use the `format` argument of the [`babel.numbers.format_decimal`][1] function, to control how many decimals you want to display, independently of the locale used.  For example:

```
>>> from babel.numbers import format_decimal
>>> format_decimal(1234.1, format='#,##0.000', locale='en_US')
u'1,234.100'
>>> format_decimal(1234.1, format='#,##0.00##', locale='en_US')
u'1,234.10'
>>> format_decimal(1234.123456, format='#,##0.00##', locale='en_US')
u'1,234.1235'
```

Also note that you can pass instances of `Decimal` (properly imported, once this PR is merged) to the number formatting functions.  And, if you are serious about decimal number precision, such as in finances, you **should** be using `Decimal` everywhere in your business logic.

In summary, this PR allows users to control the rounding behavior when digit truncation is necessary to comply with the format in use, either locale's default or custom.

[1]: http://babel.pocoo.org/en/latest/api/numbers.html#babel.numbers.format_decimal

**Comment by sublee** [comment]
@etanol Custom `format` options work well with several locales but not whole of the locales.

The `#,##0.00##` format, you presented, is proper for `en_US` or `ru_RU` because they have the same length of comma groups.  But numbers in `hi_IN` has a different comma rule.  Commas should be inserted every 2 digits except the first group.  Only the first group has 3 digits especially.  In other words, the default number format of `hi_IN` is `#,##,##0.###`.

So, if I want to set a precision by the `format` option, I should know the comma rule of the locale I will use:

```python
>>> format_number(123456789.123456789, locale='hi_IN')
'12,34,56,789.123'
>>> # The format is yours.  The precision specified exactly but commas are wrong.
>>> format_number(123456789.123456789, format='#,##0.00##', locale='hi_IN')
'123,456,789.1235'
>>> # All is okay but the format depends on the hi_IN locale.
>>> format_number(123456789.123456789, format='#,##,##0.00##', locale='hi_IN')
'12,34,56,789.1235'
```

As I said, it depends on locales instead of business logic.

**Comment by etanol** [comment]
@sublee Right, I see what you mean.  I have a couple of ideas to refactor the `babel.numbers` module and maybe expose some tuning over the number formatting by skipping over string patterns.  But due to lack of spare time, don't expect any change (at least from my part) until the end of this year.

**Comment by sublee** [comment]
@etanol Sounds great!  I'll wait your work.

**Comment by akx** [comment]
So yeah, even if this doesn't exactly fix #90, I think it's a step in a nice direction.

@jtwang, @sublee, what do you think, shall we merge?

**Comment by jtwang** [comment]
@akx - sure, I didn't spot anything worrisome. :)
